src/scanner.js

```diff
@@ -19,40 +17,37 @@ async function scanner(cfg, authorizer) {
   // for scans to take longer than for the auth service to be overloaded.
   let auth = new taskcluster.Auth({credentials: cfg.app.credentials});
 
-  // gather all identityProviderIds used in any authorizer
-  let identityProviders = authorizer.identityProviders;
-
-  // enumerate all clients for any of those identity providers
-  let clients = [];
-  for (let idp of identityProviders) {
-    clients = clients.concat(await auth.listClients({prefix: idp + '/'}));
-  }
-
-  // sort by clientId, so that each identity (a prefix of the clientId) appears
-  // contiguously
-  clients = _.sortBy(clients, 'clientId');
+  const clients = await auth.listClients();
 
   // iterate through the clients, constructing a new User as necessary, comparing
   // the client's scopes to the User's scopes and disabling where necessary.
   let user, userScopes;
   let idPattern = /^([^\/]*\/[^\/]*)\/.+$/;
-  for (let client of clients.sort()) {
+  for (let client of clients) {
     debug('examining client', client.clientId);
     if (!client.clientId.match(idPattern) || client.disabled) {
       continue;
     }
 
     // refresh the user if it does not correspond to this client
-    let clientIdentity = client.clientId.replace(idPattern, '$1');
-    if (!user || user.identity != clientIdentity) {
-      user = new User();
-      user.identity = clientIdentity;
+    let urlEncodedIdentity = client.clientId.replace(idPattern, '$1');
+
+    if (!user || user.identity != urlEncodedIdentity) {
+      await Promise.all(Object
+        .keys(cfg.handlers)
+        .map(async h => {
+          const handler = handlers[h];
+          const identity = urlEncodedIdentity.split('/', 2)[1];
+          const cleanIdentity = identity.startsWith('github') ?
+            identity.substr(0, identity.lastIndexOf(encodeURIComponent('|'))) :
+            identity;
 
-      await authorizer.authorize(user);
+          user = await handler.userFromIdentity(cleanIdentity);
+        }));
 
       userScopes = (await auth.expandScopes({scopes: user.scopes()})).scopes;
-      // allow the implicit 'assume:client-id:<clientId> auth adds for each client
-      userScopes.push('assume:client-id:' + clientIdentity + '/*');
+      // allow the implicit 'assume:client-id:<urlencodedUserId> auth adds for each client
+      userScopes.push('assume:client-id:' + urlEncodedIdentity + '/*');
```

Comment:
I think we can drop this -- it was for compatibility with an old way of doing things in the auth service.